### PR TITLE
Add dispatcher utility

### PR DESCRIPTION
**Pull request by mraspaud**
Add the dispatcher utility.

```yaml
client1:
  host: ftp://ftp.client1.com
  connection_parameters:
    credential_file: .netrc
  filepattern: '{platform_name}_{start_time}.{format}'
  directory: /input_data/{sensor}
  dispatch:
    - topics:
        - /level2/viirs
        - /level2/avhrr
      conditions:
        # key matches metadata items or provides default
        - product: [green_snow, true_color]
          sensor: viirs
        - product: [green_snow, overview]
          sensor: avhrr
          # special section "except" for negating
          except:
            platform_name: NOAA-15
    - topics:
        - /level3/cloudtype
      directory: /input/cloud_products
      conditions:
        - area: omerc_bb
          daylight: '<30'
          coverage: '>50'
```

### CODE REVIEW

**Comment by pnuu** [comment]
I'll give some more comments tomorrow morning.

File: trollmoves/dispatcher.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+      connection_parameters:
+        connection_uptime: 60
+      filepattern: '{platform_name}_{start_time}.{format}'
+      directory: /input_data/{sensor}
+      aliases:
+        product:
+          natural_color: dnc
+          overview: ovw
+      dispatch:
+        - topics:
```
Comment (pnuu) [inline trollmoves/dispatcher.py]: Is there a need to match other things, like originating host or `service`? Looking at the very generic example topics, I could see a need to pre-filter the messages with something more, e.g. when there are development streams being tested.

File: trollmoves/dispatcher.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+
+It listens to posttroll messages to find out what to dispatch.
+
+
+Format of the configuration file
+================================
+
+.. highlight:: yaml
+
+Example config::
```
Comment (pnuu) [inline trollmoves/dispatcher.py]: Could we have the example config also in the `examples/` directory?

File: trollmoves/dispatcher.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+    client1:
```
Comment (pnuu) [inline trollmoves/dispatcher.py]: I believe this is just a generic name for the configuration section? Could it be renamed to something like `target_host_1_config` to distinguish from `client` side of `trollmoves`.

File: trollmoves/dispatcher.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+      host: ftp://ftp.client1.com
```
Comment (pnuu) [inline trollmoves/dispatcher.py]: Maybe add one level more here, e.g.:
```yaml
  target_host:
    uri: ftp://ftp.client1.com
    connection_parameters:
      credential_file: .netrc
    filepattern: '{platform_name}_{start_time}.{format}'
    directory: /input_data/{sensor}
```
Just to make it more clear these will are for the target system, not e.g. `filepattern` of the incoming files.

**Comment by pnuu** [comment]
Could there be a `condition` for filename patterns? I'm thinking about messages coming from `trollmoves` transfers, which might need additional transfers to servers without `trollmoves` client capabilities, like external customers. Although now that I look at those messages I think there might be enought information even without this feature :thinking:

**Comment by mraspaud** [comment]
I was hoping to be free from the filenames and just work with metadata...

File: trollmoves/dispatcher.py (hunk 3 above)
Comment (mraspaud) [reply trollmoves/dispatcher.py]: Good point. Something like this ?
```suggestion
    target1:
```

File: trollmoves/dispatcher.py (hunk 2 above)
Comment (mraspaud) [reply trollmoves/dispatcher.py]: Yep, I'll put it there

File: trollmoves/dispatcher.py (hunk 1 above)
Comment (mraspaud) [reply trollmoves/dispatcher.py]: It's really up to the user to define what topics they would like to use. Do you want me to give a more complex example ?
Regarding service or host, for me that would be matched with the metadata somehow, but topics should work too.

File: trollmoves/dispatcher.py (hunk 4 above)
Comment (mraspaud) [reply trollmoves/dispatcher.py]: so you would like:
```yaml
client1:
  target_host:
    uri: ftp://ftp.client1.com
``` 
etc ?
I'm not sure I like it, it adds an 'unused' level, a level with just one item.

File: trollmoves/dispatcher.py (hunk 1 above)
Comment (pnuu) [reply trollmoves/dispatcher.py]: I'll be having a very detailed topics, so I'm fine with this. It was just a thought that someone might want to have only data coming from local reception to be forwarded even if there are other sources for the same products with the same topics.

File: trollmoves/dispatcher.py (hunk 3 above)
Comment (pnuu) [reply trollmoves/dispatcher.py]: Yes, that's more clear.

File: trollmoves/dispatcher.py (hunk 4 above)
Comment (pnuu) [reply trollmoves/dispatcher.py]: Yes, like that. If not, I think there should anyway be a clear separation between the message metadata bits and items relevant to the target host. Oh, if `dispatch` is renamed to `metadata_filters`, and maybe `filepattern` to `target_filepattern` and `directory` to `target_directory` that should do the trick.

**Comment by pnuu** [comment]
Isn't `filename` (or `uid`/`uri`) also metadata? :grin: Anyway, this can also be handled by having specific topics.

File: trollmoves/dispatcher.py (hunk 4 above)
Comment (mraspaud) [reply trollmoves/dispatcher.py]: `dispatch_configs` will be used as per slack conversation

**Comment by pnuu** [comment]
LGTM, feel free to merge if you don't find anything to change.

**Comment by mraspaud** [comment]
Thanks. I'll test my current dispatch chains with the modifications to make sure it still works

**Pull request merged**